File: 2ndSemester/AlgorithimsAndDataStructuresII/U09-HashTable/java/rehash/HashTable.java
public class HashTable{
	private int maxLength;
	private int [] table;
	private final int NULL = -1;

	public HashTable(){
		this(13);
	}

	public HashTable(int maxLength){
		this.maxLength = maxLength;
		this.table = new int[maxLength];
		for(int i = 0; i < maxLength; i++)
			table[i] = NULL;
	}

	public int hash(int element){
		return element % maxLength;
	}

	public int rehash(int element){
		return ++element % maxLength;
	}

	public void insert(int element){
		if(element != NULL){
			int pos = hash(element);
			if(table[pos] == NULL){
				table[pos] = element;
			}
			else{
				pos = rehash(element);
				if(table[pos] == NULL)
					table[pos] = element;
				else
					System.out.println("Sorry we are out of rehashes for this value :(");
			}
		}
	}

	public boolean search(int element){
		boolean result = false;
		int pos = hash(element);

		if(table[pos] == element || table[rehash(element)] == element)
				result = true;
		else{
			for(int i = 0; i < maxLength; i++){
				if(table[i] == element){
					result = true;
					i = maxLength;	
				}

			}
		}
		return result;
	}

	public void remove(int element){
		boolean found = false;
		if(table[hash(element)] == element || table[rehash(element)] == element){
			found = true;
			table[hash(element)] = NULL;
		}
		else {
			for(int i = 0; i < maxLength; i++){
				if(table[i] == element){
					found = true;
					table[i] = NULL;
					i = maxLength;
				}
			}
		}

		System.out.println("The element " + element + ((found)?" was removed": "does not exist on your table"));
	}
	public void display(){
		System.out.print("[");
		for(int i = 0; i < maxLength; i++)
			System.out.print(((table[i] == NULL) ? "empty" : table[i]) + ((i < table.length - 1)?", ": "]"));
	}


}

File: 2ndSemester/AlgorithimsAndDataStructuresII/U09-HashTable/c/Basic_implementation/hashtable.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/*The basic implementation of the hash table*/
#define MAX_SIZE 31

void init_table (int table[]){
    for(int i = 0; i < MAX_SIZE; i++)
        table[i] = -1;          //in this case qe assume that all values are positive or 0
}

int hash(int key){
    return key % MAX_SIZE;
}

void display(int table[]){
    int i;
    printf("Your table\n[");
    for(int i = 0; i < MAX_SIZE; i++)
        printf("\t%d = %d\n", i, table[i]);
    printf("]");
}

void insert(int table[], int key){
    int id = hash(key);
    while(table[id] >= 0)
        id = hash(++id);
    table[id] = key;
}

bool search(int table[], int key){
    int id = hash(key);
    bool result = false;
    while(table[id] > -1){
        if(table[id] == key){
            result = true;
            break;
        }

        else
            id = hash(++id);
    }
    return result;
}

void insertion_menu(int table[]){
    printf("Please insert a value:");
    int value;
    scanf(" %d", &value);
    insert(table,value);
}

void search_menu(int table[]){
    printf("Please inser a value to search: ");
    int key;
    scanf(" %d", &key);
    if(search(table,key) == true)
        printf("The element %d exists on yout table\n", key);
    else
        printf("The element %d does not exist your your table\n", key);
}

int main(){
    int option, table[MAX_SIZE];
    init_table(table);
    do{
        printf("(0)Quit\n(1)Insert\n(2)Search\n(3)Display\n");
        scanf(" %d", &option);
        switch (option)
        {
        case 0:
            printf("Thanks for using the hashtable app\n");
            break;
        case 1:
            insertion_menu(table);
            break;
        case 2:
            search_menu(table);
            break;
        case 3: 
            display(table);
            break;
        
        default:
            break;
        }
    }while(option != 0);
    return 0; 
}